Updates for Everything Blog (#2260)
IBM/FHIR: docs/src/pages/blog/everything.md

---
layout: post
title:  Introducing the IBM FHIR Server - $everything operation
description: Introducing the IBM FHIR Server - $everything operation 
date:   2021-04-21
---

# Introducing the IBM FHIR Server - $everything operation

> In [IBM FHIR Server 4.7.0](https://github.com/IBM/FHIR/releases/tag/4.7.0), the IBM FHIR Server team turned on the `Patient/$everything` operation. In [IBM FHIR Server 4.6.1](https://github.com/IBM/FHIR/releases/tag/4.6.1), an external contributor and an IBMer, [Luis García](https://github.com/luisgarcc) implemented  the operation [#1044](https://github.com/IBM/FHIR/issues/1044). 

Author: ```suggestion
In [IBM FHIR Server 4.7.0](https://github.com/IBM/FHIR/releases/tag/4.7.0), the IBM FHIR Server team turned on the `Patient/$everything` operation. We have this feature thanks to external contributor and an IBMer, [Luis García](https://github.com/luisgarcc) who implemented  the operation [#1044](https://github.com/IBM/FHIR/issues/1044). 
```

> In [IBM FHIR Server 4.7.0](https://github.com/IBM/FHIR/releases/tag/4.7.0), the IBM FHIR Server team turned on the `Patient/$everything` operation. In [IBM FHIR Server 4.6.1](https://github.com/IBM/FHIR/releases/tag/4.6.1), an external contributor and an IBMer, [Luis García](https://github.com/luisgarcc) implemented  the operation [#1044](https://github.com/IBM/FHIR/issues/1044). The core team merged the initial implementation of the `Patient/$everything` operation.

Reviewer: I'd maybe flip this so that it goes chronological.  And in either case, I think could do without the last sentence here.  Just my 2 cents.

Author: changed slightly to reflect the real intent (the thanks and attribution)

> In [IBM FHIR Server 4.7.0](https://github.com/IBM/FHIR/releases/tag/4.7.0), the IBM FHIR Server team turned on the `Patient/$everything` operation. In [IBM FHIR Server 4.6.1](https://github.com/IBM/FHIR/releases/tag/4.6.1), an external contributor and an IBMer, [Luis García](https://github.com/luisgarcc) implemented  the operation [#1044](https://github.com/IBM/FHIR/issues/1044). The core team merged the initial implementation of the `Patient/$everything` operation.

Author: ```suggestion
In [IBM FHIR Server 4.7.0](https://github.com/IBM/FHIR/releases/tag/4.7.0), the IBM FHIR Server team turned on the `Patient/$everything` operation - an external contributor and an IBMer, [Luis García](https://github.com/luisgarcc) implemented  the operation [#1044](https://github.com/IBM/FHIR/issues/1044). 
```

> So what do you get with `Patient/$everything`, well, following the HL7 FHIR R4 [specification](http://hl7.org/fhir/patient-operation-everything.html), you get an operation that returns a "searchset" bundle with the information related to a patient in the context it is invoked. In the IBM FHIR Server's case, the server returns whatever resources are related to the Patient - e.g. practitioners, medications, locations, organizations etc.  You also get support for `Patient/[id]/$everything` where the `[ID]` is the logical id specific to a patient.

Reviewer: ```suggestion
So what do you get with `Patient/$everything`? Well, following the HL7 FHIR R4 [specification](http://hl7.org/fhir/patient-operation-everything.html), you get an operation that returns a "searchset" bundle with the information related to a patient in the context it is invoked. In the IBM FHIR Server's case, the server returns whatever resources are related to the Patient - e.g. practitioners, medications, locations, organizations etc.  You also get support for `Patient/[id]/$everything` where the `[ID]` is the logical id specific to a patient.
```

| `_type` | code | One or more parameters, each containing one or more comma-delimited FHIR resource types to include in the return resources. In the absence of any specified types, the server returns all resource types | 
| `return` | Bundle | The bundle is "searchset" |

Note, currently the implementation ignores the `_count` parameter.

Now, let's prepopulate some sample data on the server using the [Integration Test data](https://github.com/IBM/FHIR/blob/main/fhir-server-test/src/test/resources/testdata/everything-operation/Antonia30_Acosta403.json).

1. Download the Sample Data

``` sh
curl -L https://raw.githubusercontent.com/IBM/FHIR/main/fhir-server-test/src/test/resources/testdata/everything-operation/Antonia30_Acosta403.json -o Antonia30_Acosta403.json
```

2. Load the Sample Data bundle to the IBM FHIR Server

``` sh
curl -k --location --request POST 'https://localhost:9443/fhir-server/api/v4' \
--header 'Content-Type: application/fhir+json' \

> --header "Authorization: Basic ${DUMMY_PASSWORD}" \

Reviewer: should we steer them toward the `-u` option instead?  its a little easier than getting the base64 just right

Author: good point - fixed

--data-binary  "@Antonia30_Acosta403.json" -o response.json
```

3. Scan the response.json for any status that is not `"status": "201"`.  For example, the status is in the family of User Request Error or Server Side Error.

4. Check the `response.json` and find the id for Patient (it should be the first one).

``` json
        {
            "response": {
                "id": "178f4a3f869-6abfe1e6-3a4a-4c9a-81a8-fb8c6263e003",
                "status": "201",
                "location": "Patient/178f4a3f869-6abfe1e6-3a4a-4c9a-81a8-fb8c6263e003/_history/1",
                "etag": "W/\"1\"",
                "lastModified": "2021-04-21T13:34:50.207684Z"
            }
        },
```

5. Check the `Patient/[id]/$everything`

* Request *
``` sh
curl -k --location --request GET 'https://localhost:9443/fhir-server/api/v4/Patient/178f4a3f869-6abfe1e6-3a4a-4c9a-81a8-fb8c6263e003/$everything' \
    --header 'Content-Type: application/fhir+json' \

>     --header "Authorization: Basic ${DUMMY_PASSWORD}"

Reviewer: should we steer them toward the `--user` (or `-u`) option instead?  its a little easier than getting the base64 just right

Author: Fixed
